service/history/api/startworkflow/api.go

```diff
@@ -41,167 +46,463 @@ import (
+		return nil, serviceerror.NewInternal("unexpected error: mutable state did not have an inflight workflow task")
+	}
+	workflowSnapshot, workflowEvents, err := mutableState.CloseTransactionAsSnapshot(
 		now,
 		workflow.TransactionPolicyActive,
 	)
 	if err != nil {
 		return nil, err
 	}
-	if len(newWorkflowEventsSeq) != 1 {
+	if len(workflowEvents) != 1 {
 		return nil, serviceerror.NewInternal("unable to create 1st event batch")
 	}
 
-	// create as brand new
-	createMode := persistence.CreateWorkflowModeBrandNew
-	prevRunID := ""
-	prevLastWriteVersion := int64(0)
-	err = workflowContext.GetContext().CreateWorkflowExecution(
+	return &creationContext{
+		workflowID:       workflowID,
+		runID:            runID,
+		workflowContext:  workflowContext,
+		workflowTaskInfo: workflowTaskInfo,
+		workflowSnapshot: workflowSnapshot,
+		workflowEvents:   workflowEvents,
+	}, nil
+}
+
+// createNewExecution creates a new "brand new" execution in the executions table.
+func (s *Starter) createNewExecution(ctx context.Context, creationCtx *creationContext) error {
+	now := s.shardCtx.GetTimeSource().Now()
+	return creationCtx.workflowContext.GetContext().CreateWorkflowExecution(
 		ctx,
 		now,
-		createMode,
-		prevRunID,
-		prevLastWriteVersion,
-		workflowContext.GetMutableState(),
-		newWorkflow,
-		newWorkflowEventsSeq,
+		persistence.CreateWorkflowModeBrandNew,
+		"", // prevRunID
+		0,  // prevLastWriteVersion
+		creationCtx.workflowContext.GetMutableState(),
+		creationCtx.workflowSnapshot,
+		creationCtx.workflowEvents,
 	)
-	if err == nil {
-		return &historyservice.StartWorkflowExecutionResponse{
-			RunId: runID.String(),
-		}, nil
-	}
+}
 
-	t, ok := err.(*persistence.CurrentWorkflowConditionFailedError)
-	if !ok {
+// handleConflict handles CurrentWorkflowConditionFailedError where a previous request with the same request ID already
+// created a workflow execution with a different run ID.
+// The history we generated above should be deleted by a background process.
+func (s *Starter) handleConflict(
+	ctx context.Context,
+	creationCtx *creationContext,
+	workflowConditionFailed *persistence.CurrentWorkflowConditionFailedError,
+) (*historyservice.StartWorkflowExecutionResponse, error) {
+	request := s.request.StartRequest
+	if workflowConditionFailed.RequestID == request.GetRequestId() {
+		return s.respondToRetriedRequest(ctx, workflowConditionFailed.RunID)
+	}
+	if err := s.verifyNamespaceActive(creationCtx, workflowConditionFailed); err != nil {
 		return nil, err
 	}
-
-	// handle CurrentWorkflowConditionFailedError
-	if t.RequestID == request.GetRequestId() {
-		return &historyservice.StartWorkflowExecutionResponse{
-			RunId: t.RunID,
-		}, nil
-		// delete history is expected here because duplicate start request will create history with different rid
+	if mutableStateInfo, err := s.applyWorkflowIDReusePolicy(ctx, workflowConditionFailed, creationCtx.runID); err != nil {
+		switch err {
+		case nil:
```

Comment:
L218 checked the err is not nil, then this `case nil` will never be used?

and I think the entire switch case should be moved into `applyWorkflowIDReusePolicy` as it's for handling error when there's prevExecutionUpdateAction.

Reply:
I missed this when refactoring, good catch.
I'm cleaning this up a bit.